### PR TITLE
Partially move cg_ssa towards using a single builder

### PR DESCRIPTION
**Pull request by bjorn3**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/94123)*

Not all codegen backends can handle hopping between blocks well. For example Cranelift requires blocks to be terminated before switching to building a new block. Rust-gpu requires a `RefCell` to allow hopping between blocks and cg_gcc currently has a buggy implementation of hopping between blocks. This PR reduces the amount of cases where cg_ssa switches between blocks before they are finished and mostly fixes the block hopping in cg_gcc. (~~only `scalar_to_backend` doesn't handle it correctly yet in cg_gcc~~ fixed that one.)

@antoyo please review the cg_gcc changes.

### CODE REVIEW

**Comment by bjorn3** [comment]
@bors try @rust-timer queue

just in case. Don't expect a lot of change though.

**Comment by bjorn3** [comment]
There is a small bug in the last commit only affecting cg_gcc. I will push a fix tomorrow.

**Comment by bjorn3** [comment]
Split out the cg_gcc changes where possible to my `cg_gcc_fix_non_singleton_builder` branch.

Ready for review.

**Review by antoyo — APPROVED** [review]
Looks good!

File: compiler/rustc_codegen_llvm/src/va_arg.rs [hunk 1]
```diff
@@ -125,53 +125,53 @@ fn emit_aapcs_va_arg<'ll, 'tcx>(
...
+    bx.br(end);
 
-    let val = end.phi(
-        layout.immediate_llvm_type(bx),
-        &[reg_value, stack_value],
-        &[in_reg.llbb(), on_stack.llbb()],
-    );
+    bx.switch_to_block(end);
+    let val =
+        bx.phi(layout.immediate_llvm_type(bx), &[reg_value, stack_value], &[bx.llbb(), bx.llbb()]);
```
Comment (tmiasko) [inline compiler/rustc_codegen_llvm/src/va_arg.rs:173]:
```suggestion
        bx.phi(layout.immediate_llvm_type(bx), &[reg_value, stack_value], &[in_reg, on_stack]);
```

Comment (bjorn3) [reply compiler/rustc_codegen_llvm/src/va_arg.rs]: Thanks!

**Comment by tmiasko** [comment]
Could you squash the fix into commit that made the change? Feel free to r+ when done.

**Comment by Mark-Simulacrum** [comment]
Likely just noise -- externs is currently a little noisy (and we haven't absorbed that noisiness into our statistical estimation yet). Seems to be due to https://github.com/rust-lang/rust/pull/93839, which makes me suspect this is related to PGO or inlining decisions of some kind. https://github.com/rust-lang/rust/pull/94373 might help.

**Comment by Srinvil2012** [comment]
Primarily benefits cg_gcc (by fixing scalar-to-backend handling) and aligns with the requirements of Cranelift and Rust-GPU.

**Pull request merged**